spectator-web-spring/src/main/java/com/netflix/spectator/controllers/MetricsController.java

```diff
@@ -0,0 +1,229 @@
+   * This is a helper function for the REST endpoint and to test against.
+   */
+  Map<String, MetricValues> encodeRegistry(
+        Registry sourceRegistry, MeasurementFilter filter) {
+    Map<String, MetricValues> metricMap = new HashMap<String, MetricValues>();
+
+    /**
+     * Spectator meters seam to group measurements. The meter name is the
+     * measurement name prefix. It seems that all the measurements within a
+     * meter instance share the same tag values as the meter.
+     * Different instances are different tag assignments, keeping the groups
+     * of measurements together.
+     *
+     * For now, we are not going to make this assumption and will just flatten
+     * out the whole measurement space. That makes fewer assumptions for now.
+     * If we make grouping assumptions we can factor out the labels into a more
+     * terse json. But the difference isnt necessarily that big and if the
+     * assumption doesnt hold, then forming the groupings will be complicated
+     * and more expensive.
+     */
+    for (Meter meter : sourceRegistry) {
+      Map<Id, List<Measurement>> collection
+          = new HashMap<Id, List<Measurement>>();
+      collectValues(collection, meter, filter);
+
+      if (!collection.isEmpty()) {
+        String kind = meterToKind(meter);
+        for (Map.Entry<Id, List<Measurement>> entry : collection.entrySet()) {
+           String entryName = entry.getKey().name();
+           MetricValues have = metricMap.get(entryName);
+           if (have == null) {
+             metricMap.put(entryName, MetricValues.make(kind, entry.getValue()));
+           } else {
+              have.addMeasurements(kind, entry.getValue());
+           }
+        }
+      }
+    }
+    return metricMap;
+  }
+
+  /**
+   * Looks up key in map, returning a specific default value if not found.
+   */
+  private static String getWithDefault(Map<String, String> map,
+                                       String key, String defaultValue) {
+    String value = map.get(key);
+    return value == null ? defaultValue : value;
+  }
+
+  /**
+   * Collect all the meter values matching the tag pattern.
+   */
+  public static List<Id> collectValues(
+      Map<Id, List<Measurement>> collection,
+      Meter meter,
+      MeasurementFilter filter) {
+    List<Id> newIds = new ArrayList<Id>();
+
+    for (Measurement measurement : meter.measure()) {
+      if (!filter.keep(meter, measurement)) {
+          continue;
+      }
+      Id id = measurement.id();
+
+      List<Measurement> valueList = collection.get(id);
+      if (valueList == null) {
+        valueList = new ArrayList<Measurement>();
+        collection.put(id, valueList);
+        newIds.add(id);
+      }
+      valueList.add(measurement);
+    }
+    return newIds;
+  }
+
+  /**
+   * Determine the type of a meter for reporting purposes.
+   */
+  public static String meterToKind(Meter meter) {
```

Comment:
Ideally Meter has a public interface I could use instead.
I think I can make this package private since I dont require it. However I think others might. I'd use it in the StackdriverWriter however because of the way Stackdriver defines a counter I need to treat everything as a gauge so its intrinsic kind turns out not to matter.